docs/source/cuda/examples.rst

```diff
@@ -29,58 +35,80 @@ they may not be large enough to hold the entire inputs at once).
-for threads in a block to cooperately compute on a task.  The following
+for threads in a block to cooperatively compute on a task.  The following
 implements a faster version of the square matrix multiplication using shared
-memory::
-
-    from numba import cuda, float32
-
-    # Controls threads per block and shared memory usage.
-    # The computation will be done on blocks of TPBxTPB elements.
-    TPB = 16
-
-    @cuda.jit
-    def fast_matmul(A, B, C):
-        # Define an array in the shared memory
-        # The size and type of the arrays must be known at compile time
-        sA = cuda.shared.array(shape=(TPB, TPB), dtype=float32)
-        sB = cuda.shared.array(shape=(TPB, TPB), dtype=float32)
-
-        x, y = cuda.grid(2)
-
-        tx = cuda.threadIdx.x
-        ty = cuda.threadIdx.y
-        bpg = cuda.gridDim.x    # blocks per grid
-
-        if x >= C.shape[0] or y >= C.shape[1]:
-            # Quit if (x, y) is outside of valid C boundary
-            return
-
-        # Each thread computes one element in the result matrix.
-        # The dot product is chunked into dot products of TPB-long vectors.
-        tmp = 0.
-        for i in range(bpg):
-            # Preload data into shared memory
-            sA[tx, ty] = A[x, ty + i * TPB]
-            sB[tx, ty] = B[tx + i * TPB, y]
-
-            # Wait until all threads finish preloading
-            cuda.syncthreads()
-
-            # Computes partial product on the shared memory
-            for j in range(TPB):
-                tmp += sA[tx, j] * sB[j, ty]
+memory:
 
-            # Wait until all threads finish computing
-            cuda.syncthreads()
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_matmul.py
+   :language: python
+   :caption: from ``test_ex_matmul`` in ``numba/cuda/tests/doc_example/test_matmul.py``
+   :start-after: magictoken.ex_fast_matmul.begin
+   :end-before: magictoken.ex_fast_matmul.end
+   :dedent: 8
+   :linenos:
 
-        C[x, y] = tmp
 
-Because the shared memory is a limited resources, the code preloads small
+Because the shared memory is a limited resource, the code preloads small
 block at a time from the input arrays.  Then, it calls
 :func:`~numba.cuda.syncthreads` to wait until all threads have finished
 preloading and before doing the computation on the shared memory.
 It synchronizes again after the computation to ensure all threads
 have finished with the data in shared memory before overwriting it
 in the next loop iteration.
 
+An example usage of this function is as follows:
+
+.. literalinclude:: ../../../numba/cuda/tests/doc_examples/test_matmul.py
+   :language: python
+   :caption: from ``test_ex_matmul`` in ``numba/cuda/tests/doc_example/test_matmul.py``
+   :start-after: magictoken.ex_run_fast_matmul.begin
+   :end-before: magictoken.ex_run_fast_matmul.end
+   :dedent: 8
+   :linenos:
+
+
+This passes a :ref:`CUDA memory check test <developer-debugging>`, which
```

Comment:
Shouldn't this link to the section on debugging CUDA Python code?

Reply:
Yes, I think you're right. I had a ref in there that didn't seem to work, but I think I need to find the one not in the developer docs.

Reply:
I've updated this to reference <cuda-debugging-python-code> which has an anchor added to the relevant section.